module/icp/algs/sha2/sha2.c

```diff
@@ -52,19 +54,138 @@
 static void Encode(uint8_t *, uint32_t *, size_t);
 static void Encode64(uint8_t *, uint64_t *, size_t);
 
+
 /* userspace only supports the generic version */
 #if	defined(__amd64) && defined(_KERNEL)
-#define	SHA512Transform(ctx, in) SHA512TransformBlocks((ctx), (in), 1)
-#define	SHA256Transform(ctx, in) SHA256TransformBlocks((ctx), (in), 1)
+
+#include <sys/simd.h>
+
+typedef void (*sha2_block_f)(SHA2_CTX *ctx, const void *in, size_t num);
 
 void SHA512TransformBlocks(SHA2_CTX *ctx, const void *in, size_t num);
 void SHA256TransformBlocks(SHA2_CTX *ctx, const void *in, size_t num);
 
 #else
 static void SHA256Transform(SHA2_CTX *, const uint8_t *);
 static void SHA512Transform(SHA2_CTX *, const uint8_t *);
+
+static void SHA512TransformBlocks(SHA2_CTX *ctx, const void *in, size_t num);
+static void SHA256TransformBlocks(SHA2_CTX *ctx, const void *in, size_t num);
+
 #endif	/* __amd64 && _KERNEL */
 
+/*
+ * For compatibility with existing code, generic and x86_64 may point to the
+ * same implementation.
+ */
+static alg_impl_ops_t sha256_impl_generic = {
+	SHA256TransformBlocks, alg_impl_will_always_work, 0, "generic"};
+
+#if defined(__x86_64)
+static alg_impl_ops_t sha256_x86_64_generic = {
+	SHA256TransformBlocks, alg_impl_will_always_work, 1, "x86_64"};
+
+#if defined(HAVE_AVX) && defined(_KERNEL)
```

Comment:
Should these optimizations also be available in userspace?

How would they be configured, since `icp_sha256_impl_set` is only compiled for kernel.

Reply:
Not available in userspace.